README.md
![logo](https://raw.githubusercontent.com/codedge-llc/pigeon/master/docs/logo.png)

> HTTP2-compliant wrapper for sending iOS and Android push notifications.

[![CI](https://github.com/codedge-llc/pigeon/actions/workflows/ci.yml/badge.svg)](https://github.com/codedge-llc/pigeon/actions/workflows/ci.yml)
[![Hex.pm](http://img.shields.io/hexpm/v/pigeon.svg)](https://hex.pm/packages/pigeon)
[![Hex.pm](http://img.shields.io/hexpm/dt/pigeon.svg)](https://hex.pm/packages/pigeon)

_Pigeon v2.0 is in release candidate status. See [the latest stable 1.6 on Hex](https://hex.pm/packages/pigeon)
or [the 1.6 branch on GitHub](https://github.com/codedge-llc/pigeon/tree/v1.6) for installation._

## Installation

Add `:pigeon` and as a `mix.exs` dependency:

```elixir
def deps do
  [
    {:pigeon, github: "mdlkxzmcp/pigeon", tag: "2.0.0-rc.2"}
  ]
end
```

## Getting Started

Check the module documentation for your push notification service.

- [Pigeon.ADM](https://hexdocs.pm/pigeon/2.0.0-rc.1/Pigeon.ADM.html) - Amazon Android.
- [Pigeon.APNS](https://hexdocs.pm/pigeon/2.0.0-rc.1/Pigeon.APNS.html) - Apple iOS.
- [Pigeon.FCM](https://hexdocs.pm/pigeon/2.0.0-rc.1/Pigeon.FCM.html) - Firebase Cloud Messaging v1 API.
- [Pigeon.LegacyFCM](https://hexdocs.pm/pigeon/2.0.0-rc.1/Pigeon.LegacyFCM.html) - Firebase Cloud Messaging Legacy API.

### Creating Dynamic Runtime Dispatchers

Pigeon can spin up dynamic dispatchers for a variety of advanced use-cases, such as
supporting dozens of dispatcher configurations or custom connection pools.

See [Pigeon.Dispatcher](https://hexdocs.pm/pigeon/2.0.0-rc.1/Pigeon.Dispatcher.html) for instructions.

### Writing a Custom Dispatcher Adapter

Want to write a Pigeon adapter for an unsupported push notification service?

See [Pigeon.Adapter](https://hexdocs.pm/pigeon/2.0.0-rc.1/Pigeon.Adapter.html) for instructions.

## Contributing

### Testing

Unit tests can be run with `mix test` or `mix coveralls.html`. Environment variables will need to be set for
various credentials. See [config/test.exs](https://github.com/codedge-llc/pigeon/blob/master/config/test.exs)
for the full list.

### Formatting

This project uses Elixir's `mix format` and [Prettier](https://prettier.io) for formatting.
Add hooks in your editor of choice to run it after a save. Be sure it respects this project's
`.formatter.exs`.

### Commits

Git commit subjects use the [Karma style](http://karma-runner.github.io/5.0/dev/git-commit-msg.html).

## License

Copyright (c) 2015-2023 Codedge LLC (https://www.codedge.io/)

This library is MIT licensed. See the [LICENSE](https://github.com/codedge-llc/pigeon/blob/master/LICENSE) for details.


CHANGELOG.md
# Changelog

## Unreleased

## v2.0.0-rc.2

**Added**
- Add ExpiredToken as an option for APNS error response
- Improve handling of FCM error responses

**Changed**
- Bump Certifi, Httpoison, Hackney, Jason and ExCoveralls

**Fixed**

- Resolve APNS `:too_many_provider_token_updates` by moving token generation into
  `APNS.Token` ([#227](https://github.com/codedge-llc/pigeon/pull/227)).
- Support HTTPoison 2.0. ([#236](https://github.com/codedge-llc/pigeon/pull/236))
- Use a fork of Kadabra to fix ssl_error
- Fix httpoison dependency not resolving 2.0
- Fix DispatcherWorker missing a clause for timeout

## v2.0.0-rc.1

**Fixed**

- Reset stream ID when connection is closed. ([#216](https://github.com/codedge-llc/pigeon/pull/217))

**Changed**

- Bump `goth` dependency to `~> 1.3.0`. ([#224](https://github.com/codedge-llc/pigeon/pull/224))
- Bump minimum supported Elixir version to `1.7`.

## v2.0.0-rc.0

**Changed**

- Default JSON library set to Jason.
  ([#182](https://github.com/codedge-llc/pigeon/pull/182))
- Pigeon application module moved from `Pigeon` to `Pigeon.Application`.
  ([#183](https://github.com/codedge-llc/pigeon/pull/183))
- Kadabra bumped to v0.5.0, and now a required dependency.
  ([#184](https://github.com/codedge-llc/pigeon/pull/184))
- Sending a list of pushes synchronously now actually sends them one at a time. For production
  workloads, using the async `:on_response` callback is strongly suggested.

**Fixed**

- Removed Elixir 1.11 compile warnings.
  ([#184](https://github.com/codedge-llc/pigeon/pull/184))

**Removed**

- `:certfile` and `:keyfile` are no longer valid options for APNS configurations.
  Instead, read the file before loading (e.g. `cert: File.read!("cert.pem")`)
  ([#183](https://github.com/codedge-llc/pigeon/pull/183))
- `:debug_log` removed.

## v1.6.0

**Added**

- JSON library made configurable. For backwards compatibility, Poison is still a required dependency.
  Override in your `config.exs`.

  ```
  config :pigeon, json_library: Jason
  ```

**Fixed**

- Handle FCM single message_id's on topic pushes.

## v1.5.1

- Added APNS InvalidPushType error ([#172](https://github.com/codedge-llc/pigeon/pull/172)).
- Fixed various typespecs ([#170](https://github.com/codedge-llc/pigeon/pull/170)).

## v1.5.0

- Bumped minimum Elixir version to 1.6
- Raise `Pigeon.ConfigError` when booting invalid config structs.
  See below for validated keys and error types.
- `APNS.JWTConfig` now validates key p8 content before connecting.
- Relaxed `gen_stage` dependency to allow `~> 1.0`

Validated config keys:

- `ADM.Config` - `:client_id`, `:client_secret`
- `APNS.Config` - `:cert`, `:key`
- `APNS.JWTConfig` - `:team_id`, `:key`, `:key_identifier`
- `FCM.Config` - `:key`

Possible error values:

- `{:error, {:invalid, value}}`
- `{:error, {:nofile, value}}`

## v1.4.0

- `apns-push-type` header support for iOS 13. An additional `:push_type` key has been
  added to the `APNS.Notification` struct.

## v1.3.2

- Document workers configuration for run-time configuration of push workers.
- Modify run-time configuration of push workers so that multiple (or no)
  workers may be returned by the startup configuration.

## v1.3.1

- Joken dependency bumped to 2.1

## v1.3.0

- Support for FCM `content_available`, `mutable_content`, and `condition` keys
- Set `priority` of APNS notifications
- Joken dependency bumped to 2.0.1

## v1.2.4

- Fixed ADM handling of connection timeouts

## v1.2.3

- Fixed APNS, FCM and ADM error response parse crashes. Error responses not
  listed in the documentation are returned as `:unknown_error`

## v1.2.2

- Fixed APNS handling of notification `expiration`
- Added APNS support for `collapse_id`

## v1.2.1

- FCM notifications can now handle `time_to_live`, `collapse_key`, `restricted_package_name`
  and `dry_run` keys

## v1.2.0

- Support for APNS JWT configuration
- Bump `kadabra` dependency to `v0.4.2`

## v1.1.6

- Relax `gen_stage` dependency to `~> 0.12`
- Bump `kadabra` dependency to `v0.3.7`

## v1.1.5

- Fix: relax `httpoison` dependency to allow `0.x` or `1.0`

## v1.1.4

- Fix: `:on_response` callbacks spawned as supervised task instead of running
  in the `Worker` process
- Fix: ADM token refresh failure returns updated notification instead of
  error tuple

## v1.1.3

- More robust FCM/APNS backpressure
- Bumped minimum Kadabra version to `v0.3.6`

## v1.1.2

- Auto-restart connections if max stream ID is reached
- FCM/APNS Workers now use GenStage to queue pending pushes
- Bumped minimum Kadabra version to `v0.3.5`

## v1.1.1

- Bumped minimum Kadabra version to `v0.3.4`

## v1.1.0

- Minimum requirements now Elixir v1.4 and OTP 19.2 (Kadabra bumped
  to `v0.3.0`)
- Startup worker configs. Create a functions that return config
  structs and specify them your `config.exs` with

```elixir
config :pigeon, workers: [
  {YourApp.Pigeon, :apns_config},
  {YourApp.Pigeon, :fcm_config},
  {YourApp.Pigeon, :adm_config},
  ...
]
```

**APNS**

- `APNS.Config.config/1` renamed to `APNS.Config.new/1`
- `APNS.push/2` tagged tuples done away with in favor of a `:response` key on
  the notification.
- Override push server endpoint with `:uri` option in `APNS.Config.new/1`
- `:use_2197` renamed to `:port`
- `:uri` config option for overriding push server endpoint
- `:reconnect` now false by default

**FCM**

- `NotificationResponse` done away with in favor of a `:response` key
  on `Notification`
- Override push server endpoint with `:uri` and `:port` options
  in `FCM.Config.new/1`
- `:uri` and `:port` config options for overriding push server endpoint

**ADM**

- `ADM.Config.config/1` renamed to `ADM.Config.new/1`
- `ADM.push/2` tagged tuples done away with in favor of a `:response` key on
  the notification.
- `ADM.start_connection/1` and `ADM.stop_connection/1` added

## v1.0.4

- Fix: removed connection pinging from FCM.Worker (`:ping_period` option
  left in FCM config to not break API)

## v1.0.3


mix.exs
defmodule Pigeon.Mixfile do
  use Mix.Project

  @source_url "https://github.com/codedge-llc/pigeon"
  @version "2.0.0-rc.2"

  def project do
    [
      app: :pigeon,
      build_embedded: Mix.env() == :prod,
      deps: deps(),
      description: description(),
      dialyzer: dialyzer(),
      docs: docs(),
      elixir: "~> 1.7",
      elixirc_options: [warnings_as_errors: true],
      elixirc_paths: elixirc_paths(Mix.env()),
      name: "Pigeon",
      package: package(),
      preferred_cli_env: [
        coveralls: :test,
        "coveralls.detail": :test,
        "coveralls.post": :test,
        "coveralls.html": :test
      ],
      source_url: @source_url,
      start_permanent: Mix.env() == :prod,
      test_coverage: [tool: ExCoveralls],
      version: @version
    ]
  end

  defp elixirc_paths(:dev), do: ["lib", "test/support"]
  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

  def application do
    [
      extra_applications: [:logger],
      mod: {Pigeon.Application, []}
    ]
  end

  defp deps do
    [
      {:credo, "~> 1.0", only: [:dev, :test], runtime: false},
      {:dialyxir, "~> 1.0", only: [:dev], runtime: false},
      {:earmark, "~> 1.0", only: :dev, runtime: false},
      {:excoveralls, "~> 0.16.1", only: :test, runtime: false},
      {:ex_doc, ">= 0.0.0", only: :dev, runtime: false},
      {:goth, "~> 1.3.0"},
      {:httpoison, "~> 1.0 or ~> 2.0"},
      {:jason, "~> 1.4", optional: true},
      {:joken, "~> 2.1"},
      {:kadabra, github: "mdlkxzmcp/kadabra", tag: "v0.7.0"}
    ]
  end

  defp docs do
    [
      groups_for_modules: [
        "ADM - Amazon Android": [Pigeon.ADM, Pigeon.ADM.Notification],
        "APNS - Apple iOS": [Pigeon.APNS, Pigeon.APNS.Notification],
        "FCM - Firebase Cloud Messaging": [
          Pigeon.FCM,
          Pigeon.FCM.Notification,
          Pigeon.LegacyFCM,
          Pigeon.LegacyFCM.Notification
        ]
      ],
      main: "Pigeon"
    ]
  end

  defp description do
    """
    HTTP2-compliant wrapper for sending iOS (APNS), Android (FCM),
    and Amazon Android (ADM) push notifications.
    """
  end

  defp package do
    [
      files: ["lib", "mix.exs", "README*", "LICENSE*"],
      licenses: ["MIT"],
      links: %{
        "Changelog" => "https://hexdocs.pm/pigeon/changelog.html",
        "GitHub" => @source_url
      },
      maintainers: ["Henry Popp", "Tyler Hurst"]
    ]
  end

  defp dialyzer do
    [
      plt_file: {:no_warn, "priv/plts/dialyzer.plt"}
    ]
  end
end
